plugins/envision_delta/main.py

```diff
@@ -0,0 +1,131 @@
+    def add_delta(self, index, year, idu, field, new_value):
+        if self.base_year is None:
+            self.base_year = year
+
+        if year - self.base_year not in self.year_index_map:
+            self.year_index_map[year - self.base_year] = index
+
+        self.deltas.append(Delta(year, idu, field, new_value))
+
+    def index_from_year(self, year):
+        if year - self.base_year in self.year_index_map:
+            return self.year_index_map[year - self.base_year]
+        return -1
+
+    def index_range_from_year(self, year):
+        return self.index_range_from_year_range(year, year + 1)
+
+    def index_range_from_year_range(self, from_year, to_year):
+        count = len(self.deltas)
+        end_year = self.deltas[count - 1].year + 1
+
+        if from_year >= end_year:
+            from_index_closed = count
+        else:
+            from_index_closed = self.index_from_year(from_year)
+
+        if to_year >= end_year:
+            to_index_open = count
+        else:
+            to_index_open = self.index_from_year(to_year)
+
+        if to_year < from_year:
+            from_index_closed -= 1
+            to_index_open -= 1
+
+        return from_index_closed, to_index_open
+
+    @property
+    def timestamps(self):
+        return sorted([datetime.datetime(year=year + self.base_year, month=1, day=1) for year in self.year_index_map])
+
+
+class EnvisionDeltaArray(ArrayDataPlugin):
+
+    id = 'envision_delta_reader'
+    name = 'Envision Delta Array Data Plugin'
+    description = 'Loads ENVISION delta array data from a CSV file.'
+    author = 'Conservation Biology Institute'
+    version = '1.0'
+    extensions = [('csv', 'CSV')]
+
+    data_name = None
+    time_info = None
+    variables = None
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.data_name = 'Envision Delta Array'
+        self.time_info = TemporalInfo()
+        self.delta_array = DeltaArray()
+        self.variables = []
+
+    def load_data(self):
+        self.variables.clear()
+        self.delta_array.deltas.clear()
+        self.delta_array.year_index_map.clear()
+        self.data_name = self.path.split(os.sep)[-1].split('.')[0]
+
+        with open(self.path, 'r') as f:
+            reader = csv.DictReader(f)
+            for idx, row in enumerate(reader):
+                field = row.get('field')
+                if field not in self.variables:
+                    self.variables.append(field)
+                self.delta_array.add_delta(idx, int(row.get('year')), int(row.get('idu')), field,
+                                           float(row.get('newValue')))
+
+        timestamps = self.delta_array.timestamps
+        if timestamps:
+            self.time_info.timestamps = timestamps
+
+    @staticmethod
+    def is_valid_file(path):
+        return True     # Todo - assert that time is formatted correctly
+
+    #def calculate_stats(self): # Todo - calculate stats for delta arrays
```

Comment:
Does this need to be commented out?

Reply:
Ah, it should be implemented. I'll do that!